Language/Haskell/GhcMod/Monad/Types.hs

```diff
@@ -129,10 +138,11 @@ data GhcModState = GhcModState {
       gmGhcSession   :: !(Maybe GmGhcSession)
     , gmComponents   :: !(Map ChComponentName (GmComponent GMCResolved (Set ModulePath)))
     , gmCompilerMode :: !CompilerMode
+    , gmMMappedFiles :: Map TargetId (StringBuffer, UTCTime)
```

Comment:
Maybe I didn't explain this properly. The problem is that guessTarget can return either a module name or a file name but you use Eq/Ord on the TargetId type when finding out if a target is mapped. So if guessTarget returns a module name like say `Language.GhcMod.Target` but then the user/editor frontend passes ghc-mod the filename of that module the mapping wont apply.

So you need to find a way to either map module names onto absolute paths or paths onto module names. Usually the former is done using the ModuleGraph from ghc's `depanal` but since depanal won't honor the file remapping that might not be an option.

So instead you can modify [HomeModuleGraph](https://github.com/kazu-yamamoto/ghc-mod/blob/master/Language/Haskell/GhcMod/HomeModuleGraph.hs) to also honor your mapping. Right now HMG is not meant to be absolutely correct since it's "only" used to find out which modules/files belong to which cabal components. So errors coming from it are mostly ignored since we'll still get a partial module graph that might work. The infrastructure for logging and getting out errors and warnings is there though just needs to be actually used.

HMG is fairly forgiving already, it only parses the "module head" i.e. `module X where` and import statements any parse errors after which will not disturb it. Only processors are a major pain and will need special handling.